### PR TITLE
users: create or modify: add support for external IDs

### PR DESCRIPTION
**Pull request by rissson**
Also includes a small fix in case the return password in null.

### CODE REVIEW

File: src/users/create_or_modify/v2.rs [hunk 1]
```diff
@@ -90,3 +96,9 @@ impl Response {
         Self { details }
     }
 }
+
+/// An external ID associated with a user
+// TODO
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct IncomingExternalId {}
```
Comment (jplatte) [inline src/users/create_or_modify/v2.rs:104]: This should not be needed. Did you previously have an `ExternalId` type with a lifetime parameter?

Comment (rissson) [reply src/users/create_or_modify/v2.rs]: Yes indeed, that's my bad, I removed it.

File: src/users.rs [hunk 2]
```diff
@@ -62,4 +62,17 @@ pub struct UserDetails {
...
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub external_ids: Vec<ExternalId>,
+}
+
+/// An external ID associated with a user
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ExternalId {
+    /// The authentication provider to which the user is associated.
+    pub auth_provider: String,
+    /// The ID known to the auth provider associated with this user.
```
Comment (jplatte) [inline src/users.rs:76]:
```suggestion
    pub auth_provider: String,

    /// The ID known to the auth provider associated with this user.
```

File: src/users/create_or_modify/v2.rs [hunk 3]
```diff
@@ -34,6 +34,10 @@ ruma_api! {
         #[serde(skip_serializing_if = "Option::is_none")]
         pub threepids: Option<Vec<ThirdPartyIdentifier>>,
 
+        /// defaults to empty, or the current value if user already exists
```
Comment (jplatte) [inline src/users/create_or_modify/v2.rs:37]:
```suggestion
        /// Defaults to empty, or the current value if user already exists
```

File: src/users.rs [hunk 4]
```diff
@@ -18,7 +18,7 @@ pub struct UserDetails {
     pub name: String,
 
     /// The password hash of th account
```
Comment (jplatte) [inline src/users.rs:20]:
```suggestion
    /// The password hash of the account.
```

**Review by jplatte — COMMENTED** [review]
Some nitpicks

**Comment by jplatte** [comment]
Can I ask what you are using this crate for? 🙂

**Comment by rissson** [comment]
I use it for https://github.com/rissson/spacehub-rs, which is a tool that syncs your Matrix spaces with a folder structure, and adds/removes users based on their LDAP groups (which is destined to be an optional feature). It's not yet ready, but watch TWIM for a first release :D

**Pull request merged**